src/main/java/graphql/util/Traverser.java

```diff
@@ -0,0 +1,139 @@
+     * @param visitor a Visitor object to be notified during traversal
+     * @return      some data produced by the last Visitor's method invoked
+     */
+    public <U> Object traverse (T root, U data, TraverserVisitor<? super T, ? super U> visitor) {
+        return traverse(Collections.singleton(root), data, visitor);
+    }
+    
+    
+    /**
+     * Starts traversal of a tree from a provided roots using specified Visitor 
+     * and initial data to pass around Visitor's methods
+     * 
+     * @param <U>   type of data argument to Visitor's methods
+     * @param roots  multiple subtree roots to start traversal from
+     * @param data  some data to pass across Visitor's methods. Visitor's methods
+     * can change that data to some other values of the same type or special Traverser
+     * markers {@link TraverserMarkers}
+     * @param visitor a Visitor object to be notified during traversal
+     * @return      some data produced by the last Visitor's method invoked
+     */
+    public <U> Object traverse (Collection<T> roots, U data, TraverserVisitor<? super T, ? super U> visitor) {
+        Objects.requireNonNull(roots);
+        Objects.requireNonNull(visitor);
+        
+        stack.addAll(roots);
+        
+        Object d = data;
+        while (!(stack.isEmpty() || (d = traverseOne((TraverserVisitor<T, U>)visitor, (U)d)) == TraverserMarkers.QUIT));
+        
+        return d;
+    }
+    
+    protected <U> Object traverseOne (TraverserVisitor<T, U> visitor, U data) {
+        TraverserContext<T> top = stack.pop();
+        
+        Object result;
+        if (top == TraverserMarkers.END_LIST) {
+            // end-of-list marker, we are done recursing children, 
+            // mark the current node as fully visited
+            result = visitor.leave(stack.pop(), data);
+        } else if (top == TraverserMarkers.MAP_KEY) {
+            // in case we've traversed through a map of children,
+            // this is a chance to introspect the current child's map key
+            result = visitor.mapKey(stack.pop(), data);
+        } else if (top.isVisited(data)) {
+            // cyclic reference detected
+            result = visitor.backRef(top, data);
+        } else if ((result = visitor.enter(top, data)) == TraverserMarkers.ABORT) {
+            // abort traversing this subtree, don't recurse to children
+        } else if (result == TraverserMarkers.QUIT) {
+            // complete abort traversing, don't recurse to children
+        } else {
+            // recurse to children.
+            // Depending on RecursionState implementation we'll either
+            // put children into a stack and this will be a depth-first search
+            // -or-
+            // put children into a queue and this will be a breadth-first search
+            stack.pushAll(top, getChildren);
+        }        
+        
+        return result;
+    }
+    
+    protected final RecursionState<T> stack;
```

Comment:
please don't use protected variables and please put them on on the top of the class